### DISCUSSION TITLE
Use generators instead of array methods in data processing pipelines

### DISCUSSION THREAD

**Discussion by ikusteu** [librocco/librocco#250] [Ideas]
_"Premature optimisation is the root of all evil"_

Having said that...we might want to optimise our data processing pipelines (here I'm talking about note/warehouse `entries`, but it's applicable in general):
- we're currently using a lot of `Array.prototype.<method>`s for iterations over our data
- for a lot of data, we stream the entire snapshot on each change
- this means that for each transmission (each change in the db that causes the new value to be emitted to an rxjs pipeline) we make a new copy of an array/object for each step in the rxjs pipeline

Iterable control flows are found to be much more efficient then array methods (`for .. of` vs `Array.prototype.forEach`). Additionally, using generators, we can compose the iterations, e.g.:
```ts
// Example using array methods 
function processArr<T>(arr: T[]) {
  // Iterate over array to map it, creating a new array
  return arr.map(...)
    // Iterate over the new array to filter, returning another new array
    .filter(...)
    // Iterate over the new array to map, returning yet another new array
    .map(...)
    // Iterate over the filtered array to reduce it
    .reduce(...)
}

// Example using generators
function processIter(iterable: Iterable<T>) {
  // Create a generator, which didn't iterate over 'iterable'
  // but knows how to iterate over it and what to do with every element
  const mapped1 = mapIter(iterable, ...)
  // Create a new generator which did't iterate over 'mapped1' generator
  // but knows how to iterate over it and what to do with every element
  const filtered = filterIter(mapped1, ...)
  // ...
  const mapped2 = mapIter(filtered, ...)
  // Consume the 'mapped2' generator by iterating over all of the values
  // of the 'iterable' and applying transformations of the entire 'pipeline' on
  // each element
  const reduced = reduceIter(mapped2, ...)
}
```

The functions over iterables are implemented using generator functions under the hood, something like this:

```ts
function* mapIter<T, R>(iterable: Iterable<T>, mapper: (v: T) => R): Iterable<R> {
    for (const el of iterable) {
        yield mapper(el)
    }
}
```

Calling `mapIter` on an iterable () returns a generator:

```ts
// Remember, arrays also satisfy the iterable interface
const arr = [0, 1, 2]
// The returned 'mapped' is a generator
const mapped = mapIter(arr, (v) => v + 1) 
```

A generator is a special kind of iterable, which contains only the instructions on how to produce the next value. In this case, our generator keeps reference to the underlaying `iterable` (passed as argument) and saves only the instructions how to produce the next value, based on the next value of the underlaying `iterable` (in this case - add 1 to each element and yield the result).

One caveat of a generator is that it's single use, meaning: after we iterate over it, it's exhausted and can't be iterated over again:

```ts
const generator = mapIter(someIterable, someMapper)
const arr = [...generator] // Creates an array by iterating over the generator and stores all yelded values to the array
const arr2 = [...generator] // Creates an empty array as the generator has already been exhausted

To remedy this, if we expect the resulting generator to be reused, we can return a generator function instead, like so:

```ts
function mapIter<T, R>(iterable: Iterable<T>, mapper: (v: T) => R): () => Iterable<R> {
    return function* () {
        for (const el of iterable) {
            yield mapper(el)
        }
    }
}
```

This way, if we need to reuse the mapping logic, we can do it like so:

```ts
const arr = [0, 1, 2]
const arrMapper = mapIter(arr, (v) => v + 1)
// arrMapper is a generator function, to get the generator, we simply call it
const mapped = arrMapper()

// Both of the following arrays are the same - they contain all of the values resulting from iterating over the
// arrMapper-created generator
const mappedArray = [...mapped]
const mappedArray2 = [...arrMapper()]
```

All in all, the two main benefits of using generators for data processing are:
- iterable control flow (`for .. of`) is much faster than `Array.prototype.<method>` (sometimes even twice as fast)
- the memory footprint of a function doing the processing is much smaller, especially when multiple operations are chained together because generators, until they're consumed, save only the instruction on how to process each element of the iteration and reference to the original iterable, rather than generating a fully new array (for instance) for each step in the processing pipeline

Last but not least: Svelte's `{#each .. as}` control flow supports iterables (and is not constrained to only arrays), so we could, if we so chose, pass this logic all the way down to the UI (rather than passing copies of arrays at each step).

As conclusion: Using genrators might improve performance (both in terms of speed and memory), it does introduce additional complexity (albeit a managable one). Whether or not this ticket is worked on should really depend on whether or not we need said performance boosts (especially considering #249 will probably give us much more performance by design)

**Comment by silviot** [comment]
I completely agree with this plan. In particular I believe it's important not to copy data that can be kept unique. It takes time and space to copy it, so avoiding that has immediate benefits that compound as the app becomes more complex.

**Comment by ikusteu** [comment]
The more I work on the db, it's data processing pipelines and structures, I would like to break up some larger helpers into smaller ones and be able to mix and match them. With current implementation, that's possible, but would produce for much inefficiency (inefficiency which will grow with the complexity of data processing).

With respect to that, I think it makes sense to give this proposal a go. This can be done in phases (to split up the work), so I've opened up the following tickets:

- **Phase 1** - #257 
- **Phase 2** - #258 
- **Phase 3** - #259
